objectstore: Reporting predicted available space

aclamk opened the pull request:
The PR addresses problems related to reporting available space in objectstores.

We need:
- a fixed value attached to device that we use for crush weight
   It is done by reinforcing documentation for `store_statfs_t.total`. 
- a dynamic value reflecting store capacity for user data
   Done by adding `store_statfs_t.est_capacity` that predicts capacity based on current usage.

In addition: 
 - Added `store_statfs_t.est_available`, a dynamic value reflecting how much more user data can objectstore still accomodate.

Replaces: https://github.com/ceph/ceph/pull/66342

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

ifed01 committed:
osd,mon,extblkdev: Use store_statfs_t::*_raw members for tracking physical device stats

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>
(cherry picked from commit 34aff5c5f9e3f3b224fa2f28dde1e9ecca32d65e)
PR: https://github.com/ceph/ceph/pull/63526

aclamk committed:
os/bluestore: Implement estimates in store_statfs_t

Add code for calculation of est_capacity and est_available.
BlueStore compressed store can now attempt to approximate estimated capacity
and estimated available space.

Signed-off-by: Adam Kupczyk <akupczyk@ibm.com>

aclamk committed:
os/bluestore: Implement estimates in store_statfs_t for ExtBlkDev

Add code for calculation of est_capacity and est_available for ExtBlkDev plugins.
The algorithm uses device physical and logical sizes to approximate estimated capacity
and estimated available space.

Signed-off-by: Adam Kupczyk <akupczyk@ibm.com>

aclamk committed:
osd: Expanding store_statfs_t fields, documentation

Add documentation for the fields. Clarify their meaning.
Add 2 new fields that track:
- estimated capacity that store is likely to achieve
- estimated available that reflects how much more user data can be
  accomodated

Signed-off-by: Adam Kupczyk <akupczyk@ibm.com>

aclamk committed:
osd/OSDMap.cc: Refactor OSDUtilizationDumper

Created struct common_stats_t that agregates stat values.
Now dump_item functions are much nicer.

Signed-off-by: Adam Kupczyk <akupczyk@ibm.com>

bill-scales reviewed (approved):
LGTM. I wonder whether the 0.1% and 1% limits used when estimating capacity are the best choice - especially for the small capacity FCM where you don't have to write very much data to reach these limits, but I don't have any better suggestions and I don't think it matters if estimated capacity fluctuates a bit when the system is virtually empty

aclamk committed:
osd: Expanding store_statfs_t fields

Adapt test to also expect:
- total_estimated_capacity
- total_estimated_avail_bytes

Signed-off-by: Adam Kupczyk <akupczyk@ibm.com>
